src/keymap-manager.coffee

```diff
@@ -409,26 +409,51 @@ class KeymapManager
   # target is `.defaultTarget` if that property is assigned on the keymap.
   #
   # * `event` A `KeyboardEvent` of type 'keydown'
-  handleKeyboardEvent: (event) ->
+  handleKeyboardEvent: (event, isReplay) ->
+    eventResult = @_handleKeyboardEvent(event, @queuedKeystrokes)
+    return unless eventResult?
+    {keystroke, queuedKeystrokes, exactMatch, partialMatches} = eventResult
+
+    hasPartialMatches = partialMatches? and partialMatches.length > 0
+
+    @queuedKeystrokes.push(keystroke)
+    @queuedKeyboardEvents.push(event)
+
+    if queuedKeystrokes? and hasPartialMatches
+      enableTimeout = (
+        @pendingStateTimeoutHandle? or
+        exactMatch? or
+        characterForKeyboardEvent(@queuedKeyboardEvents[0])?
+      )
+      enableTimeout = false if isReplay
+      @enterPendingState(partialMatches, enableTimeout)
+    else if not exactMatch? and not hasPartialMatches and @pendingPartialMatches?
+      @terminatePendingState()
+    else
+      @clearQueuedKeystrokes()
+
+  # This function does not set any state. State is handled in the
+  # `handleKeyboardEvent` function.
+  _handleKeyboardEvent: (event, queuedKeystrokes=[]) ->
```

Comment:
@nathansobo how do you feel about this naming? Have another idea? I like them split into 2 functions as it lends clarity IMO, as the state is set in a pretty straightforward way in the public function, and this does the actual handling. But we dont use `_` functions anywhere.